### PR TITLE
Update AcquireTokenSilent api to handle single accounts

### PR DESCRIPTION
**Pull request by trwalke**
Implementation of #1180 
currently in progress but very close to completion

### CODE REVIEW

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenSilentParameterBuilder.cs [hunk 1]
```diff
@@ -56,6 +63,12 @@ private AcquireTokenSilentParameterBuilder WithLoginHint(string loginHint)
             return this;
         }
 
+        private AcquireTokenSilentParameterBuilder WithSingleAccount()
+        {
+            //CommonParameters.AddApiTelemetryFeature(ApiTelemetryFeature.WithLoginHint);
```
Comment (trwalke) [inline src/Microsoft.Identity.Client/ApiConfig/AcquireTokenSilentParameterBuilder.cs]: I am assuming we want to add another telemetry entry here? @henrik-me @jmprieur

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 2]
```diff
@@ -183,7 +215,7 @@ private RequestContext CreateRequestContext(Guid telemetryCorrelationId)
         /// specify extra query parameters
         ///
         /// </remarks>
-        public AcquireTokenSilentParameterBuilder AcquireTokenSilent(IEnumerable<string> scopes, IAccount account)
+        public AcquireTokenSilentParameterBuilder AcquireTokenSilentWithAccount(IEnumerable<string> scopes, IAccount account)
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/ClientApplicationBase.cs]: >AcquireTokenSilentWithAccount [](start = 50, length = 29)

I thought we agreed not to change this to ensure we don't break people? If we go on with this proper migration aid should be provided ensuring the compiler will say exactly what needs to be done.

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 3]
```diff
@@ -219,7 +251,7 @@ public AcquireTokenSilentParameterBuilder AcquireTokenSilent(IEnumerable<string>
         /// specify extra query parameters
         ///
         /// </remarks>
-        public AcquireTokenSilentParameterBuilder AcquireTokenSilent(IEnumerable<string> scopes, string loginHint)
+        public AcquireTokenSilentParameterBuilder AcquireTokenSilentWithLoginHint(IEnumerable<string> scopes, string loginHint)
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/ClientApplicationBase.cs]: >AcquireTokenSilentWithLoginHint [](start = 50, length = 31)

Same comment as above. perhaps we should add deprecation messages and not completely remove? Though given we move to v4 perhaps ok. Lots of samples needs to be updated based on these changes.

**Comment by henrik-me** [comment]
> using System;

same comments as for the implementation

---
Refers to: src/Microsoft.Identity.Client/IClientApplicationBase.cs:4 in c47dbae. [](commit_id = c47dbaead4389dc250c34a604b5ac98bfbc8453e, deletion_comment = False)

File: src/Microsoft.Identity.Client/ApiConfig/AcquireTokenSilentParameterBuilder.cs (hunk 1 above)
Comment (henrik-me) [reply src/Microsoft.Identity.Client/ApiConfig/AcquireTokenSilentParameterBuilder.cs]: makes sense yes.

---
In reply to: [288809223](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1181#discussion_r288809223) [](ancestors = 288809223)

File: src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs [hunk 4]
```diff
@@ -29,16 +29,19 @@ internal class SilentRequest : RequestBase
             _silentParameters = silentParameters;
         }
 
-        private async Task<IAccount> GetSingleAccountForLoginHintAsync(string loginHint)
+        private async Task<IAccount> GetSingleAccountForLoginHintAsync(string loginHint, bool singleAccount = false)
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: do you need to add this? don't you have this in the _silentParameters?

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ClientCredentialWithCertTest.cs [hunk 5]
```diff
@@ -203,9 +203,8 @@ public async Task JsonWebTokenWithX509PublicCertSendCertificateSilentTestAsync()
                 harness.HttpManager.AddMockHandler(CreateTokenResponseHttpHandlerWithX5CValidation(false));
 
                 var result = await app
-                    .AcquireTokenSilent(
-                        new[] { "someTestScope"},
-                        new Account(MsalTestConstants.UserIdentifier, MsalTestConstants.DisplayableId, null))
+                    .AquireTokenSilentWithSingleAccount(
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ClientCredentialWithCertTest.cs]: >AquireTokenSilentWithSingleAccount [](start = 21, length = 34)

not sure this was the original intention of the test, but suppose it's an ok way of validating

**Comment by henrik-me** [comment]
One test seems to have been adjusted to fit the added method. It should be considered if adding some specific tests to validate error conditions should be added as well, also the positive tests should probably be added as well though some test have been re-used to fit the need.

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 2 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Client/ClientApplicationBase.cs]: I second Henrik. I thought we don't want to change the name and signature of existing methods. I would really not want us to break anybody here. just bring a bit of comfort.

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 3 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Client/ClientApplicationBase.cs]: We cannot release a new version 3 weeks after the GA and require everybody to update absolutely all their code.

File: src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs (hunk 4 above)
Comment (trwalke) [reply src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: not needed. its dead code

File: tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ClientCredentialWithCertTest.cs (hunk 5 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/ClientCredentialWithCertTest.cs]: not entirely but it does make the test simpler. an additional unit test for AquireTokenSilentWithSingleAccount() will be added

**Pull request closed without merging**